### PR TITLE
Remote device access, editing and snapshotting with Auth enabled

### PR DESCRIPTION
**Pull request by Steve-Mcl**
## Description

Support remote device access, editing and snapshotting with NR token Auth to prevent local access

### TODO:

- [x] documentation
- [x] tests
- [ ] audit log entries - open a task to capture this

### Notes for debugging and comprehension

1. The auth method is essentially as described in the issue mermaid diagram except for the route names as described by Bens comment in the following comment n the issue
2. Search the code for "Enable Device Editor" to see the 12 steps of acheiving a successful tunnel 
   1. image
   2. Each step indicates source and destination, protocol and a brief description
   3. This will also aid completing documentation
3. I stand on the shoulders of giants - Ben. The tunnelling was first realised by Ben (thanks @hardillb) BUT adding Auth and controlling the tunnel, reigning in the many of async problems (had to add cmnd/resp support!), head twisting, torturous job to get it all working (to prevent local access but permit tunnelled access). So my apologies for the jumbled tired comments, lateness of this PR and the size of it but after working ~48h SOLID (Fri ~ Now) and debugging 401s, 503s, xhr cancelled, no response, no auto-login, wrong token, wrong auth header, missing auth headers, alternative auth methods and a thousand other problems, I am really really t...
 

## Related PRs

https://github.com/flowforge/flowforge-device-agent/pull/77

## Related Issue(s)

#1821 

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/flowforge/flowforge/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [x] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `flowforge/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `flowforge/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Backport needed? -> add the `backport` label
 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
Following dev-demo live (p)review, additional pushes have been made to address ui/ux and security.

#### Security
* URL for Admin is /device-editor/  ✅
* access_token=xxxx appended to editor url  ✅
* FF Platform account necessary to access  (browser should have session) 
   * I.e. only operable when logged in to FF ✅
   * I.e. not operable from another browser using proxy URL or local access ✅
#### Enterprise Only 
   * Elements are hidden for community edition ✅
#### UI / UX
   * Developer Options - add smaller, lighter “Developer Mode Only” text to the right ✅
   * Device Editor should be 3 columns  ✅
      * Label
      * State
      * Toggle Button
   * Create Snapshot should be 3 columns ✅
      * Label
      * Blank
      * Create Snapshot button
   * Tooltip for the flask icon 
      * Should say “Developer Mode” ✅
   * Device Mode Modal 
      * Remove mention of “Autonomous mode” from UI - essentially, Developer mode on/off ✅


Still todo (either this or later iteration)
* Timeout of enabling tunnel should revert state to show actual state
   * We witnessed the device being "broken" but tunnel still showed "enabled". State should reflect condition.
   * NOTE: This is slightly different now we only have a "toggle" button for enabling/disabling device editor access

**Comment by knolleary** [comment]
> Timeout of enabling tunnel should revert state to show actual state
We witnessed the device being "broken" but tunnel still showed "enabled". State should reflect condition.
NOTE: This is slightly different now we only have a "toggle" button for enabling/disabling device editor access

I think we definitely need something in the first iteration then ensures the state doesn't get out of sync if the device is offline. Otherwise it will say the editor is enabled, but the button never comes active - without any indication what is wrong.

**Comment by Steve-Mcl** [comment]
New Tests passing (except the common `Reached heap limit Allocation failed - JavaScript heap out of memory`)

**Comment by knolleary** [comment]
Is there a summary of all of the new api routes this PR adds? It's quite hard to get the big picture trying to pick them all out individually and the external apis are a key part of the design.

PRs of this size and complexity should ideally summarise their key changes in the description, including db changes, api changes etc.

Given this is EE licensed functionality, some of this should be under the `ee` src tree - such as the `DeviceTunnelManager` and `api/deviceEditor` - so the relevant routes only get added if a license is applied.
 
> Timeout of enabling tunnel should revert state to show actual state
We witnessed the device being "broken" but tunnel still showed "enabled". State should reflect condition.
NOTE: This is slightly different now we only have a "toggle" button for enabling/disabling device editor access

Has this been resolved in the subsequent commits?

**Comment by Steve-Mcl** [comment]
@knolleary 

I have merged and performed local testing (approx 1hr, enable, disable, with MQTT down, with device up/down etc)
Much more solid with `sendOrQueue` mod to agent. (thanks for that)


fixed up button text wrapping in [f82ad24](https://github.com/flowforge/flowforge/pull/2042/commits/f82ad246d7949ae31668b41a4faa34fbf4c55265)

chrome_1WShKzm9vF


Other than that, ready for review/merge.

**Review by knolleary — APPROVED** [review]

File: docs/user/devices.md [hunk 1]
```diff
@@ -156,30 +156,46 @@ backup of the Node-RED flows and configuration.
 This model allows you to develop your flows in FlowForge and only push it out
 to the registered devices when you're happy with what you've created.
 
-## Editing the Node-RED flows directly on a device
+## Editing the Node-RED flows on a device
```
Comment (Steve-Mcl) [inline docs/user/devices.md:159]: docs update title "Editing the Node-RED flows directly on a device" -> "Editing the Node-RED flows on a device" will break link in `packages/flowforge/docs/user/snapshots.md`
need to change link to `See [Working with Devices](./devices.md#editing-the-node-red-flows-on-a-device) for more information.`

Comment (Steve-Mcl) [reply docs/user/devices.md:159]: I'm guessing `Test Documentation Links / test (pull_request)` doesnt check internal MD links?

File: forge/routes/api/device.js [hunk 2]
```diff
@@ -357,11 +359,79 @@ module.exports = async function (app) {
...
+            if (app.comms) {
+                app.comms.devices.sendCommandToProjectDevices(request.project.Team.hashid, request.project.id, 'update', {
+                    snapshot: snapShot.hashid
+                })
+            }
+        }
+        reply.send(app.db.views.ProjectSnapshot.snapshot(snapShot))
+    })
+
+    async function assignDeviceToProject (device, project) {
```
Comment (Steve-Mcl) [inline forge/routes/api/device.js:429]:
```suggestion
    // #region Helpers
    async function assignDeviceToProject (device, project) {
```

Missing `#region`

**Review by knolleary — APPROVED** [review]

**Pull request merged**